modules/core/src/va_intel.cpp

```diff
@@ -0,0 +1,517 @@
+
+        cl_command_queue q = (cl_command_queue)Queue::getDefault().ptr();
+
+        cl_mem images[2] = { clImageY, clImageUV };
+        status = clEnqueueAcquireVA_APIMediaSurfacesINTEL(q, 2, images, 0, NULL, NULL);
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clEnqueueAcquireVA_APIMediaSurfacesINTEL failed");
+        if (!ocl::ocl_convert_bgr_to_nv12(clBuffer, (int)u.step[0], u.cols, u.rows, clImageY, clImageUV))
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: ocl_convert_bgr_to_nv12 failed");
+        clEnqueueReleaseVA_APIMediaSurfacesINTEL(q, 2, images, 0, NULL, NULL);
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clEnqueueReleaseVA_APIMediaSurfacesINTEL failed");
+
+        status = clFinish(q); // TODO Use events
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clFinish failed");
+
+        status = clReleaseMemObject(clImageY); // TODO RAII
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clReleaseMem failed (Y plane)");
+        status = clReleaseMemObject(clImageUV);
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clReleaseMem failed (UV plane)");
+    }
+    else
+# endif // HAVE_VA_INTEL && HAVE_OPENCL
+    {
+        VAStatus status = 0;
+
+        status = vaSyncSurface(display, surface);
+        if (status != VA_STATUS_SUCCESS)
+            CV_Error(cv::Error::StsError, "VA-API: vaSyncSurface failed");
+
+        VAImage image;
+        status = vaDeriveImage(display, surface, &image);
+        if (status != VA_STATUS_SUCCESS)
+            CV_Error(cv::Error::StsError, "VA-API: vaDeriveImage failed");
+
+        unsigned char* buffer = 0;
+        status = vaMapBuffer(display, image.buf, (void **)&buffer);
+        if (status != VA_STATUS_SUCCESS)
+            CV_Error(cv::Error::StsError, "VA-API: vaMapBuffer failed");
+
+        CV_Assert(image.format.fourcc == VA_FOURCC_NV12);
+
+        Mat m = u.getMat(ACCESS_READ);
+        copy_convert_bgr_to_nv12(image, m, buffer);
+
+        status = vaUnmapBuffer(display, image.buf);
+        if (status != VA_STATUS_SUCCESS)
+            CV_Error(cv::Error::StsError, "VA-API: vaUnmapBuffer failed");
+
+        status = vaDestroyImage(display, image.image_id);
+        if (status != VA_STATUS_SUCCESS)
+            CV_Error(cv::Error::StsError, "VA-API: vaDestroyImage failed");
+    }
+#endif  // !HAVE_VA
+}
+
+void convertFromVASurface(VADisplay display, VASurfaceID surface, Size size, OutputArray dst)
+{
+    (void)display; (void)surface; (void)dst; (void)size;
+#if !defined(HAVE_VA)
+    NO_VA_SUPPORT_ERROR;
+#else  // !HAVE_VA
+    const int dtype = CV_8UC3;
+
+    // TODO Need to specify ACCESS_WRITE here somehow to prevent useless data copying!
```

Comment:
There are redundant copy in case of dst=Mat only:
Mat m;
convertFromVASurface(..., m)
{
    // resize Mat m (via API OutputArray)
    // u = m.getUMat(u) in case of dGPU only: copy m data into UMat u (we will rewrite this data below completely). In case of iGPU it will try use USE_HOST_MEM which may allow zero copy.
}